File: ai-service/app/agents/agent_loop.py
from __future__ import annotations

import json
import time
from typing import Any

from app.agents.state import AgentLoopConfig, AgentLoopResult, TerminationReason
from app.repositories import agent_repository
from app.services.llm_client import get_llm_client
from app.tools.base import ToolContext, ToolDefinition, to_llm_tool_spec


def run_agent_loop(config: AgentLoopConfig) -> AgentLoopResult:
    deadline = time.time() + config.timeout_seconds
    tool_specs = [to_llm_tool_spec(t) for t in config.tools]

    messages: list[dict[str, Any]] = [
        {"role": "system", "content": config.system_prompt},
        {"role": "user",   "content": config.user_message},
    ]

    tool_results: dict[str, Any] = {}
    seq = 0
    tool_call_count = 0
    llm = get_llm_client()

    while True:
        # ── Hard limits ──────────────────────────────────────────────────────────
        if time.time() > deadline:
            return AgentLoopResult(tool_results, seq, tool_call_count, TerminationReason.TIMEOUT)
        if seq >= config.max_steps:
            return AgentLoopResult(tool_results, seq, tool_call_count, TerminationReason.MAX_STEPS)
        if tool_call_count >= config.max_tool_calls:
            return AgentLoopResult(tool_results, seq, tool_call_count, TerminationReason.MAX_TOOL_CALLS)

        # ── Call LLM ─────────────────────────────────────────────────────────────
        seq += 1
        llm_seq = seq
        llm_start = time.time()

        try:
            action = llm.chat_complete_with_tools(messages, tool_specs)
        except Exception as e:
            agent_repository.upsert_completed_step(
                config.agent_run_id, llm_seq, "LLM", None,
                json.dumps({"msgCount": len(messages)}), None,
                "FAILED", "LLM_ERROR", str(e),
                int((time.time() - llm_start) * 1000),
            )
            return AgentLoopResult(tool_results, seq, tool_call_count, TerminationReason.LLM_ERROR)

        llm_ms = int((time.time() - llm_start) * 1000)

        # ── Final answer ─────────────────────────────────────────────────────────
        if action.get("type") == "final_answer":
            agent_repository.upsert_completed_step(
                config.agent_run_id, llm_seq, "LLM", None,
                json.dumps({"msgCount": len(messages)}),
                json.dumps({"content": action.get("content", "")}),
                "COMPLETED", None, None, llm_ms,
            )
            break

        # ── Unexpected response ──────────────────────────────────────────────────
        if action.get("type") != "tool_call" or not action.get("tool_name"):
            agent_repository.upsert_completed_step(
                config.agent_run_id, llm_seq, "LLM", None,
                json.dumps({"msgCount": len(messages)}),
                json.dumps({"raw": action}),
                "FAILED", "INVALID_ACTION", "LLM returned unexpected action type", llm_ms,
            )
            break

        tool_name = action["tool_name"]
        tool_args = action.get("tool_args") or {}

        # ── LLM decided to call a tool ────────────────────────────────────────────
        agent_repository.upsert_completed_step(
            config.agent_run_id, llm_seq, "LLM", tool_name,
            json.dumps({"msgCount": len(messages), "args": tool_args}),
            json.dumps({"decided": tool_name}),
            "COMPLETED", None, None, llm_ms,
        )

        tool: ToolDefinition | None = next(
            (t for t in config.tools if t.name == tool_name), None
        )
        if not tool:
            messages.append({"role": "assistant", "content": f"[unknown tool: {tool_name}]"})
            messages.append({"role": "tool", "name": tool_name, "content": json.dumps({"error": "Tool not found"})})
            continue

        # ── Scope check ───────────────────────────────────────────────────────────
        if tool.requires_student_scope:
            req_student_id = tool_args.get("studentId")
            if req_student_id and req_student_id != config.student_id:
                seq += 1
                agent_repository.upsert_completed_step(
                    config.agent_run_id, seq, "TOOL", tool.name,
                    json.dumps(tool_args), None,
                    "FAILED", "SCOPE_VIOLATION", "Cannot access another student's data", 0,
                )
                return AgentLoopResult(tool_results, seq, tool_call_count, TerminationReason.SCOPE_VIOLATION)

        # ── Execute tool ──────────────────────────────────────────────────────────
        seq += 1
        tool_seq = seq
        tool_call_count += 1

        agent_repository.insert_running_tool_step(
            config.agent_run_id, tool_seq, tool.name, json.dumps(tool_args)
        )

        tool_start = time.time()
        ctx = ToolContext(student_id=config.student_id, agent_run_id=config.agent_run_id)
        result = tool.execute(tool_args, ctx)
        tool_ms = int((time.time() - tool_start) * 1000)

        if result.success:
            agent_repository.update_tool_step(
                config.agent_run_id, tool_seq,
                json.dumps(result.data), "COMPLETED", None, None, tool_ms,
            )
            tool_results[tool.name] = result.data
            messages.append({"role": "assistant", "content": f"[called: {tool.name}]"})
            messages.append({"role": "tool", "name": tool.name, "content": json.dumps(result.data)})
        else:
            agent_repository.update_tool_step(
                config.agent_run_id, tool_seq,
                None, "FAILED",
                result.error_code or "TOOL_ERROR",
                result.error_message or "Unknown tool error",
                tool_ms,
            )
            messages.append({"role": "assistant", "content": f"[called: {tool.name}]"})
            messages.append({"role": "tool", "name": tool.name, "content": json.dumps({"error": result.error_message})})

    return AgentLoopResult(tool_results, seq, tool_call_count, TerminationReason.NATURAL)

File: backend/src/agents/agentLoop.ts
import axios from 'axios';
import { db } from '../shared/db/pool';
import { ToolDefinition, ToolContext, findTool, toLLMToolSpec } from './tools';
import { env } from '../config/env';

// ── Public types ──────────────────────────────────────────────────────────────

export interface AgentLoopConfig {
  agentRunId: string;
  studentId: string;
  systemPrompt: string;
  userMessage: string;
  tools: ToolDefinition[];
  maxSteps: number;
  maxToolCalls: number;
  timeoutMs: number;
}

export type TerminationReason =
  | 'NATURAL'
  | 'MAX_STEPS'
  | 'MAX_TOOL_CALLS'
  | 'TIMEOUT'
  | 'LLM_ERROR'
  | 'SCOPE_VIOLATION';

export interface AgentLoopResult {
  toolResults: Record<string, unknown>;
  stepCount: number;
  toolCallCount: number;
  terminationReason: TerminationReason;
}

// ── Internal types ────────────────────────────────────────────────────────────

interface LLMMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string;
  name?: string;
}

interface LLMAction {
  type: 'tool_call' | 'final_answer';
  tool_name?: string;
  tool_args?: Record<string, unknown>;
  content?: string;
}

// ── Agent loop ────────────────────────────────────────────────────────────────

export async function runAgentLoop(config: AgentLoopConfig): Promise<AgentLoopResult> {
  const { agentRunId, studentId, maxSteps, maxToolCalls, timeoutMs } = config;
  const deadline  = Date.now() + timeoutMs;
  const toolSpecs = config.tools.map(toLLMToolSpec);

  const messages: LLMMessage[] = [
    { role: 'system', content: config.systemPrompt },
    { role: 'user',   content: config.userMessage },
  ];

  const toolResults: Record<string, unknown> = {};
  let seq          = 0;
  let toolCallCount = 0;

  while (true) {
    // ── Hard limits — checked at the top of every iteration ──────────────────
    if (Date.now() > deadline) {
      return { toolResults, stepCount: seq, toolCallCount, terminationReason: 'TIMEOUT' };
    }
    if (seq >= maxSteps) {
      return { toolResults, stepCount: seq, toolCallCount, terminationReason: 'MAX_STEPS' };
    }
    if (toolCallCount >= maxToolCalls) {
      return { toolResults, stepCount: seq, toolCallCount, terminationReason: 'MAX_TOOL_CALLS' };
    }

    // ── Ask the LLM what to do next ───────────────────────────────────────────
    const llmSeq   = ++seq;
    const llmStart = Date.now();
    let action: LLMAction;

    try {
      const resp = await axios.post<LLMAction>(
        `${env.AI_SERVICE_URL}/learning/chat-complete`,
        { messages, tools: toolSpecs },
        { timeout: 30_000 }
      );
      action = resp.data;
    } catch (e) {
      const msg = e instanceof Error ? e.message : 'LLM call failed';
      await insertCompletedStep(
        agentRunId, llmSeq, 'LLM', null,
        { msgCount: messages.length }, null,
        'FAILED', 'LLM_ERROR', msg, Date.now() - llmStart
      );
      return { toolResults, stepCount: seq, toolCallCount, terminationReason: 'LLM_ERROR' };
    }

    const llmMs = Date.now() - llmStart;

    // ── Final answer — natural termination ────────────────────────────────────
    if (action.type === 'final_answer') {
      await insertCompletedStep(
        agentRunId, llmSeq, 'LLM', null,
        { msgCount: messages.length }, { content: action.content ?? '' },
        'COMPLETED', null, null, llmMs
      );
      break;
    }

    // ── Unexpected response ───────────────────────────────────────────────────
    if (action.type !== 'tool_call' || !action.tool_name) {
      await insertCompletedStep(
        agentRunId, llmSeq, 'LLM', null,
        { msgCount: messages.length }, { raw: action },
        'FAILED', 'INVALID_ACTION', 'LLM returned unexpected action type', llmMs
      );
      break;
    }

    // ── LLM decided to call a tool ────────────────────────────────────────────
    await insertCompletedStep(
      agentRunId, llmSeq, 'LLM', action.tool_name,
      { msgCount: messages.length, args: action.tool_args ?? {} },
      { decided: action.tool_name },
      'COMPLETED', null, null, llmMs
    );

    const tool = findTool(config.tools, action.tool_name);
    if (!tool) {
      messages.push({ role: 'assistant', content: `[unknown tool: ${action.tool_name}]` });
      messages.push({ role: 'tool', name: action.tool_name, content: JSON.stringify({ error: 'Tool not found' }) });
      continue;
    }

    // ── Authorization check — in code, never trust the LLM ───────────────────
    if (tool.authorization.requiresStudentScope) {
      const reqId = (action.tool_args as Record<string, unknown> | undefined)?.studentId as string | undefined;
      if (reqId && reqId !== studentId) {
        const toolSeq = ++seq;
        await insertCompletedStep(
          agentRunId, toolSeq, 'TOOL', tool.name,
          action.tool_args ?? {}, null,
          'FAILED', 'SCOPE_VIOLATION', "Cannot access another student's data", 0
        );
        return { toolResults, stepCount: seq, toolCallCount, terminationReason: 'SCOPE_VIOLATION' };
      }
    }

    // ── Execute the tool ──────────────────────────────────────────────────────
    const toolSeq = ++seq;
    toolCallCount++;

    await db.query(
      `INSERT INTO agent.agent_steps
         (agent_run_id, sequence_no, step_type, tool_name, input, status)
       VALUES ($1,$2,'TOOL',$3,$4,'RUNNING')`,
      [agentRunId, toolSeq, tool.name, JSON.stringify(action.tool_args ?? {})]
    );

    const toolStart = Date.now();
    const ctx: ToolContext = { studentId, agentRunId };
    const result = await tool.execute(action.tool_args ?? {}, ctx);
    const toolMs = Date.now() - toolStart;

    if (result.success) {
      await db.query(
        `UPDATE agent.agent_steps
         SET output=$3, status='COMPLETED', duration_ms=$4
         WHERE agent_run_id=$1 AND sequence_no=$2`,
        [agentRunId, toolSeq, JSON.stringify(result.data), toolMs]
      );
      toolResults[tool.name] = result.data;
      messages.push({ role: 'assistant', content: `[called: ${tool.name}]` });
      messages.push({ role: 'tool', name: tool.name, content: JSON.stringify(result.data) });
    } else {
      await db.query(
        `UPDATE agent.agent_steps
         SET status='FAILED', error_code=$3, error_message=$4, duration_ms=$5
         WHERE agent_run_id=$1 AND sequence_no=$2`,
        [
          agentRunId, toolSeq,
          result.errorCode  ?? 'TOOL_ERROR',
          result.errorMessage ?? 'Unknown tool error',
          toolMs,
        ]
      );
      messages.push({ role: 'assistant', content: `[called: ${tool.name}]` });
      messages.push({ role: 'tool', name: tool.name, content: JSON.stringify({ error: result.errorMessage }) });
    }
  }

  return { toolResults, stepCount: seq, toolCallCount, terminationReason: 'NATURAL' };
}

// ── Step persistence helper ───────────────────────────────────────────────────

async function insertCompletedStep(
  runId: string,
  seqNo: number,
  stepType: string,
  toolName: string | null,
  input: unknown,
  output: unknown,
  status: string,
  errorCode: string | null,
  errorMessage: string | null,
  durationMs: number
): Promise<void> {
  await db.query(
    `INSERT INTO agent.agent_steps
       (agent_run_id, sequence_no, step_type, tool_name, input, output,
        status, error_code, error_message, duration_ms)
     VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)
     ON CONFLICT (agent_run_id, sequence_no) DO NOTHING`,
    [
      runId, seqNo, stepType, toolName,
      JSON.stringify(input ?? null),
      output !== null ? JSON.stringify(output) : null,
      status, errorCode, errorMessage, durationMs,
    ]
  );
}
